### PR TITLE
Enable "Run Query in New Tab" in SQL Lab

### LINKED ISSUE #1147: Run query in new tab on SQL Lab / Query History doesn't work
On Query History tab on every query there are 4 buttons:
- visualize
- overwrite text
- run query in new tab
- remove from log

Option "run query in new tab" does nothing.

https://github.com/airbnb/caravel/blob/e8088d5c9a522db5e0e165988fba47893337fd5b/caravel/assets/javascripts/SqlLab/components/QueryTable.jsx#L89

### PR DESCRIPTION
**Pull request by nickbarnwell**
Fixes #1147 – saw this was on the roadmap but it's important for the workflow I personally use so worst case I'll abandon this patch if there's already a fix in mainline.

Commit 7d33c8f shows the first approach I took, before reading a bit more about Redux; I can rebase it away and add tests if the approach in f3326a3 looks good.

n.b. This was my first time working with React/Redux, apologies in advance if this isn't idiomatic; I tried my best to base it on the rest of the codebase.

### CODE REVIEW

File: caravel/assets/javascripts/SqlLab/reducers.js [hunk 1]
```diff
@@ -34,6 +34,23 @@ export const sqlLabReducer = function (state, action) {
...
+    [actions.CLONE_QUERY_TO_NEW_TAB]() {
+      const progenitor = state.queryEditors.find((qe) =>
+          qe.latestQueryId === action.query.id);
+      const qe = {
+        id: shortid.generate(),
+        title: `Copy of ${progenitor.title}`,
+        dbId: (action.query.dbId) ? action.query.dbId : null,
+        schema: (action.query.schema) ? action.query.schema : null,
+        autorun: true,
+        sql: `${action.query.sql}`,
```
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/reducers.js]: same as  `sql: action.query.sql,`

**Review by mistercrunch — CHANGES_REQUESTED** [review]
One minor nit, but otherwise LGTM. I'm glad to see that people out there are already using SQL Lab. Thank you for the PR.

**Comment by nickbarnwell** [comment]
Thanks for the quick feedback! Definitely should've caught that nit, cargo-culted a little too hard there.

Will add tests tonight; since I was already in there I added a few for other reducer actions around QueryEditors - would you prefer those be separated out into another PR?

File: caravel/assets/javascripts/SqlLab/reducers.js [hunk 2]
```diff
@@ -34,6 +34,20 @@ export const sqlLabReducer = function (state, action) {
...
+      const qe = {
+        id: shortid.generate(),
+        title: `Copy of ${progenitor.title}`,
+        dbId: (action.query.dbId) ? action.query.dbId : null,
+        schema: (action.query.schema) ? action.query.schema : null,
+        autorun: true,
+        sql: action.query.sql,
+      };
+
+      return sqlLabReducer(state, actions.addQueryEditor(qe));
```
Comment (nickbarnwell) [inline caravel/assets/javascripts/SqlLab/reducers.js]: Wasn't super DRY essentially copying addQueryEditor into this action. This seems better, but not terribly clean either…suggestions for improvement welcome!

File: caravel/assets/spec/javascripts/sqllab/reducers_spec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import * as r from '../../../javascripts/SqlLab/reducers'
+import * as actions from '../../../javascripts/SqlLab/actions';
+import { describe, it } from 'mocha';
+import { expect } from 'chai';
+
+describe('sqlLabReducer', () => {
```
Comment (nickbarnwell) [inline caravel/assets/spec/javascripts/sqllab/reducers_spec.js]: Couldn't find anything to model these tests off of, let me know how you'd prefer them to be structured.

File: caravel/assets/spec/javascripts/sqllab/reducers_spec.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      expect(newState.queryEditors).have.length(2);
+      //expect(newState.queryEditors.filter((qe) => qe.sql == testQuery.SQL).length).to.equal(2)
+    });
+  
+    it('should have the same SQL as the cloned query', () => {
+      expect(newState.queryEditors[1].sql).to.equal(testQuery.sql);
+    });
+
+    it('should prefix the new tab title with "Copy of"', () => {
+      expect(newState.queryEditors[1].title).to.include("Copy of");
```
Comment (nickbarnwell) [inline caravel/assets/spec/javascripts/sqllab/reducers_spec.js]: Assuming that our new queryEditor will always be the last element is all well and good, but seems terribly hacky to me. Any suggestions or are you fine with it?

File: caravel/assets/spec/javascripts/sqllab/reducers_spec.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+describe('sqlLabReducer', () => {
+  describe('CLONE_QUERY_TO_NEW_TAB', () => {
+    const testQuery = { sql: "SELECT * FROM...", dbId: 1, id: 1 };
+    const state = Object.assign(r.initialState, { queries: [ testQuery ] });
+    const newState = r.sqlLabReducer(state, actions.cloneQueryToNewTab(testQuery))
+
+    it('should have at most one more tab', () => {
+      expect(newState.queryEditors).have.length(2);
+      //expect(newState.queryEditors.filter((qe) => qe.sql == testQuery.SQL).length).to.equal(2)
```
Comment (nickbarnwell) [inline caravel/assets/spec/javascripts/sqllab/reducers_spec.js]: Woops. Will remove this when I clean up any other nits from code review.

**Review by nickbarnwell — COMMENTED** [review]
There are a few places where I wasn't sure how you prefer things to be tested and/or implemented. Called them out to save you some work when reviewing.

**Comment by mistercrunch** [comment]
Thanks for the PR, this looks good to me as is, I'll do a blitz soon to get some test coverage on SQL Lab, but for now this is good!

**Pull request merged**